### ISSUE TITLE
Tool flags Zsh's "alternat syntax for complex commands" as invalid.

### ISSUE THREAD

**Issue by questionmarkexclamationpoint** [z-shell/zsh-lint#206]
Some (all?) syntaxes from [Zshmisc: Alternate forms for complex commands](https://zsh.sourceforge.io/Doc/Release/Shell-Grammar.html#Alternate-Forms-For-Complex-Commands-1) are flagged as errors. Examples:

```zsh
for i in {1..$x}; do
  if (( i + 69 > 420)) continue # << This line here, not the for
  # ... do real work
done
```

I find the above incredibly useful and far more readable than `(( i + 69 <= 420 )) || continue`, which will endlessly trip you up if you use `errreturn/errexit` (looking like these will be way more usable in 5.10 but, `(( i + 69 > 420 )) && continue` will still err(return/exit) even then, while `if (( ... )) continue` will not).

I also particularly like the form
```zsh
if (( x == y )) { success_condition } \
else { failure_condition }
```
it is almost as succinct as `[[ x == y ]] && success_condition || failure_condition`, but without the obvious control flow issue if `success_condition` can, itself fail, and trigger your failure condition unintentionally. On top of that, I find it more readable.

Some people also just prefer this syntax. to each their own.
```zsh
if [[ -o ignorebraces ]] {
  print -r -- 'ignorebraces is set'
} elif [[ -o kshglob ]] {
  print -r -- 'kshglob is set'
} else {
  print -r -- 'neither option set'
}
```

```zsh
local -a to_filter=()
for i in {$#to_filter..1}; (( i > 100 )) || to_filter[i]=()

local -a pending_tasks()
for t in $pending_tasks; consume_task t

forearch value ($someArray); do_something $value
```

i find this useful from time to time, the same construct exists for while and is similarly useful
```zsh
until (( success )) retry && success=$REPLY
```

I rarely find use for repeat, but when I do, it's something like this
```zsh
repeat 3; print hi
```

I don't know if I've ever found a better usage for select than this, avoiding hella nesting
```zsh
select option in first second third fourth; break
case $option in
  first|second) ... ;;
  third) ... ;;
  fourth) ... ;;
esac
```

I find this absolutely abhorrent but wanted to flag that it is valid zsh
```zsh
function killme; { print please god why have you forsaken me }
```

---

Full of personal preferences, of course, but a set of very valid, and very useful idioms arise from these grammars, that people use in myriad ways.

I am always always always dying for a useful holistic and circumspect zsh linter, and I am grateful for people with the drive to finally make one worth its salt. I have some suggestions for hints/rules, as well, but at the moment I cannot use the software as is. Tyty

**Comment by ss-o** [comment]
Thank you for this report. A catalogue like this, written by someone who actually uses these forms daily, is exactly the kind of input that moves the parser forward, and your point about `if (( ... )) continue` versus `(( ... )) || continue` under `errreturn` is a good argument for treating the short forms as first-class rather than as style curiosities.

We verified every form against three oracles: native `zsh -n`, the current zsh-lint `main`, and the raw parser front end with no zsh-lint compatibility layer. Results:

**Already supported**

- `if [[ ... ]] { } elif [[ ... ]] { } else { }` (your ignorebraces example) parses with a correct tree.
- `for x (list) { body }` parses with a correct tree.

**Local defects zsh-lint owns**

- `if (( x == y )) { a } \` followed by `else { b }` on the next line fails only because of the line continuation. The same line without `\` parses. The reported error position is also wrong (it points at a line that does not exist in your file). Both are bugs in our brace-form `if` support and will be tracked and fixed separately.
- `repeat 3; print hi` currently parses without error but as two unrelated commands, so rules would analyze a wrong tree. That is worse than an error. `repeat 3; do ...; done` and `repeat 3 { ... }` fail to parse. This gets its own tracked item.

**Not yet supported, confirmed valid Zsh**

- `if cond cmd` short form (`if (( i > 420 )) continue`, `if [[ -n $x ]] print yes`)
- `for x in ...; cmd`, `for x (...) cmd`, `while cond cmd`, `until cond cmd`
- `select o in a b; break`
- `function name; { ... }`

Each of these will be split into its own tracked parser-gap item so it carries a minimized fixture and acceptance criteria. Links will follow in this thread.

**Two examples that native Zsh rejects**

- `foreach value ($someArray); do_something $value` fails `zsh -n`. `foreach` needs a `{ ... }` body or an `end` terminator; the `end` form is a separate gap we will track.
- `local -a pending_tasks()` is a syntax error; `local -a pending_tasks=()` was presumably intended.

**How you can shape zsh-lint**

Your rule and hint suggestions are welcome, and the fastest way to get them evaluated:

- Parser gaps: one issue per language feature, with the smallest script that reproduces it and passes `zsh -n`. The process we follow is in `docs/project/parser-gap-workflow.md`; a minimized fixture in an issue is often most of the work.
- Rule ideas: the `Rule proposal` issue template in this repository. Proposals with corpus evidence (real scripts where the rule would fire) and a note on likely false positives are reviewed first; `docs/project/rule-policy.md` describes what a first-wave rule needs.
- Real-world code: pointing us at public Zsh code that fails to parse is as valuable as a fix. Every gap listed above is one we had not yet seen in the corpus we survey.
- Pull requests are welcome for any of the above. Parser adapters have a strict contract (exact error gating, byte-preserving retry, restored positions) described in the same workflow document, so opening the issue first saves a round trip.

The user guide lives at https://wiki.zshell.dev/community/zsh_lint. Please keep the reports coming.

**Comment by ss-o** [comment]
Tracked items, as promised:

- #207: `if ... { } \` newline `else { }` chain and the wrong error position
- #208: `repeat` (silent misparse of the short form, plus `do`/brace forms)
- #210: `if cond cmd` short form
- #211: `for name in list; cmd`, `for name (list) cmd`, `while cond cmd`, `until cond cmd`
- #212: `select name in list; cmd`
- #213: `function name; { list }`
- #214: `foreach ... end`

#209 is a separate gap found while verifying this report. Each item carries its minimized fixture and acceptance criteria; feel free to add real-world sites you hit to any of them.

**Comment by ss-o** [comment]
Survey follow-up (2026-09-16): the workspace survey after #216 exposed one more production from the same manual section, `for (( [expr1] ; [expr2] ; [expr3] )) sublist`, tracked as #241 and split from this issue like #211. The same run also exposed #242 (bare subscript key `ZI[bkp-.]`), which is a subscript gap rather than an alternate form and is not a child of this issue.

Children of this issue: #210, #211, #212, #241.

**Comment by ss-o** [comment]
Closing: every tracked child of this report is now on `main`.

| Form | Issue | Landed in |
| --- | --- | --- |
| `if list sublist` short form (`if (( i > 420 )) continue`) | #210 | #295 |
| `for name in list; cmd`, `for name (list) cmd`, `while cond cmd`, `until cond cmd` | #211 | #325 |
| `select name in list; cmd` | #212 | #305 |
| `for (( expr1 ; expr2 ; expr3 )) sublist`, brace body and single command | #241 | #326 |
| `if ... { } \` newline `else { }` chain and its error position | #207 | earlier |
| `repeat count sublist`, `do` and brace forms | #208 | #272, #308 |
| `function name; { list }` | #213 | #227 |

Every example in the report that native Zsh accepts now parses to the same tree as its `do ... done` or `then ... fi` form, with original positions, so rules see one body once. The two native-invalid examples (`foreach value ($someArray); do_something $value` and `local -a pending_tasks()`) stay rejected, as `zsh -n` rejects them.

Still open, adjacent, tracked separately:

- #214: `foreach name (list) ... end`.
- #261: a `{ list }` body after a terminated `for` list (`for i in a b; { ... }`; `for i (a b); { ... }` is the same gap). The unterminated `for i (a b) { ... }` already parses.
- #324: a multi-name `for` with a parenthesized list (`for key value (...) { ... }`).
- #327: an empty body after a `while`, `until` or `for` header (`while true print x`, which native Zsh reads as the condition `true print x` with no body, `while (( 1 ))` alone, `for i in a b;`). Found while verifying #211.

Thank you again for the catalogue; it turned into seven parser fixes.

**Comment by questionmarkexclamationpoint** [comment]
My apologies! Yes, `local -a pending tasks=()` was missing the `=` as you assumed. And, my mistake on the foreach syntax. : ) ty!
